ACM-23467 MRA delete feature

fxiang1 opened the pull request:
# 📝 Summary

**Ticket Summary (Title):**  
MRA delete feature
- Added finalizer logic to modify/clean up ClusterPermissions when MRAs are deleted
- Add managedcluster and clusterpermission to the scheme
- Make e2e tests work with podman (we're not allowed to use Docker)
- Add allows-errors label to e2e tests due to resource conflict retry logic
- Filter out e2e tests from sonar

Limitations
- During mra deletion, removing the finalizer triggers another reconcile

**Ticket Link:**  
https://issues.redhat.com/browse/ACM-23467

**Type of Change:**  
<!-- Select one -->
- [ ] 🐞 Bug Fix  
- [X] ✨ Feature  
- [ ] 🔧 Refactor
- [ ] 💸 Tech Debt
- [ ] 🧪 Test-related  
- [ ] 📄 Docs

---

## ✅ Checklist

### General

- [ ] PR title follows the convention (e.g. `ACM-12340 Fix bug with...`)
- [ ] Code builds and runs locally without errors
- [ ] No test logs/printing output, commented-out code, or unnecessary files
- [ ] All commits are meaningful and well-labeled

#### If Feature

- [ ] All acceptance criteria met
- [ ] Unit test coverage added or updated
- [ ] Relevant documentation or comments included

#### If Bugfix

- [ ] Root cause and fix summary are documented in the ticket (for future reference / errata)
- [ ] Fix tested thoroughly and resolves the issue
- [ ] Test(s) added to prevent regression

---

### 🗒️ Notes for Reviewers
<!-- Optional: anything reviewers should know, special context, etc. -->

mshort55 commented on internal/controller/multiclusterroleassignment_controller.go:
```diff
@@ -186,6 +189,46 @@ func (r *MulticlusterRoleAssignmentReconciler) Reconcile(ctx context.Context, re
+		log.Error(err, "Failed to refresh MulticlusterRoleAssignment after finalizer handling")
+		return ctrl.Result{RequeueAfter: DefaultRequeueDelay}, err
+	}
+
```
```suggestion

	for _, condition := range mra.Status.Conditions {
		if condition.Type == ConditionTypeReady && condition.ObservedGeneration == mra.Generation {
			log.Info("Spec unchanged, Ready condition generation matches. Skipping reconcile.")
			return ctrl.Result{}, nil
		}
	}
```

mshort55 replied:
Adding so we skip reconcile when spec does not change, like status updates.

mshort55 commented on internal/controller/multiclusterroleassignment_controller.go:
```diff
@@ -186,6 +189,46 @@ func (r *MulticlusterRoleAssignmentReconciler) Reconcile(ctx context.Context, re
+		// Add finalizer for create/update
+		if !controllerutil.ContainsFinalizer(&mra, FinalizerName) {
+			result := controllerutil.AddFinalizer(&mra, FinalizerName)
+			log.Info("Add finalizer ", "finalizer", FinalizerName, "result", result)
+			if err := r.Update(ctx, &mra); err != nil {
+				log.Error(err, "Failed to update MulticlusterRoleAssignment with finalizer")
+				return ctrl.Result{RequeueAfter: DefaultRequeueDelay}, err
+			}
```
```suggestion
			log.Info("Add finalizer and skip reconcile", "finalizer", FinalizerName, "result", result)
			if err := r.Update(ctx, &mra); err != nil {
				log.Error(err, "Failed to update MulticlusterRoleAssignment with finalizer")
				return ctrl.Result{RequeueAfter: DefaultRequeueDelay}, err
			}
			return ctrl.Result{}, nil
```

mshort55 replied:
With this, we can return after the finalizer is added and avoid extra reconciles. It will reconcile immediately after and take care of the spec changes.

For this to work, we also need to remove this line in the SetupWithManager method:
`		WithEventFilter(predicate.GenerationChangedPredicate{}).
`

mshort55 commented:
@fxiang1 when I ran e2e tests I saw this status update retry happening frequently:
`
2025-09-16T02:29:18Z    INFO    Status update conflict, retrying        {"controller": "multiclusterroleassignment", "controllerGroup": "rbac.open-cluster-management.io", "controllerKind": "MulticlusterRoleAssignment", "MulticlusterRoleAssignment": {"name":"test-multicluster-role-assignment-single-clusterrolebinding","namespace":"open-cluster-management-global-set"}, "namespace": "open-cluster-management-global-set", "name": "test-multicluster-role-assignment-single-clusterrolebinding", "reconcileID": "22f04ef9-f316-4e23-8478-16b98311a8a3", "error": "Operation cannot be fulfilled on multiclusterroleassignments.rbac.open-cluster-management.io \"test-multicluster-role-assignment-single-clusterrolebinding\": the object has been modified; please apply your changes to the latest version and try again", "resource": "test-multicluster-role-assignment-single-clusterrolebinding"}
`

There is a problem this predicate that I implemented in one of my previous PRs:
`
		WithEventFilter(predicate.GenerationChangedPredicate{}).
`

By removing this, and by returning after adding the finalizer (which will reconcile afterwards for the spec change), we don't get any of the retry errors.

I did some research and from what I read, we should return after adding the finalizer. I am sure you tried that originally, but the predicate filter I added caused you issues I assume. With these changes, we don't need to ignore errors in the logs in the e2e tests.

Let me know what you think.

mshort55 commented on internal/controller/multiclusterroleassignment_controller_test.go:
```diff
@@ -363,7 +383,7 @@ var _ = Describe("MulticlusterRoleAssignment Controller", Ordered, func() {
 
 		Describe("setRoleAssignmentStatus", func() {
-			It("Should add new role assignment status when not present", func() {
+			It("Should add new role assignment status when not present", Label("allows-errors"), func() {
```
"allows-errors" label is only for e2e tests, these should have no effect in unit tests

mshort55 commented on internal/controller/multiclusterroleassignment_controller_test.go:
```diff
@@ -375,7 +395,7 @@ var _ = Describe("MulticlusterRoleAssignment Controller", Ordered, func() {
 			})
 
-			It("Should update existing role assignment status", func() {
+			It("Should update existing role assignment status", Label("allows-errors"), func() {
```
"allows-errors" label is only for e2e tests, these should have no effect in unit tests

fxiang1 commented:
/assign @mshort55 

> @fxiang1 when I ran e2e tests I saw this status update retry happening frequently: `2025-09-16T02:29:18Z INFO Status update conflict, retrying {"controller": "multiclusterroleassignment", "controllerGroup": "rbac.open-cluster-management.io", "controllerKind": "MulticlusterRoleAssignment", "MulticlusterRoleAssignment": {"name":"test-multicluster-role-assignment-single-clusterrolebinding","namespace":"open-cluster-management-global-set"}, "namespace": "open-cluster-management-global-set", "name": "test-multicluster-role-assignment-single-clusterrolebinding", "reconcileID": "22f04ef9-f316-4e23-8478-16b98311a8a3", "error": "Operation cannot be fulfilled on multiclusterroleassignments.rbac.open-cluster-management.io \"test-multicluster-role-assignment-single-clusterrolebinding\": the object has been modified; please apply your changes to the latest version and try again", "resource": "test-multicluster-role-assignment-single-clusterrolebinding"}`
> 
> There is a problem with this predicate that I implemented in one of my previous PRs: `WithEventFilter(predicate.GenerationChangedPredicate{}).`
> 
> By removing this, and by returning from the reconcile loop after adding the finalizer (which will reconcile afterwards for the spec change), we don't get any of the retry errors for status updates.
> 
> I did some research and from what I read, we should return after adding the finalizer. I am sure you tried that originally, but the predicate filter I added caused issues with this. With these changes, we don't need to ignore errors in the logs in the e2e tests.
> 
> See my proposed code changes and let me know what you think.

Cool! Thanks @mshort55 ! I made all the suggested changes.

mshort55 commented:
/lgtm
